api/oc_ri.c

```diff
@@ -1562,68 +1422,423 @@ oc_ri_invoke_coap_entity_handler(coap_make_response_ctx_t *ctx,
+    // query matches the device id
+    if (oc_endpoint_is_multicast(endpoint) &&
+        !oc_ri_filter_request_by_device_id(
+          endpoint->device, ctx->preparsed_request_obj->uri_query.data,
+          ctx->preparsed_request_obj->uri_query.length)) {
+      coap_set_global_status_code(CLEAR_TRANSACTION);
+      coap_set_status_code(ctx->response, CLEAR_TRANSACTION);
+      return false;
     }
-    success = true;
+
+    request_obj.query = ctx->preparsed_request_obj->uri_query.data;
+    request_obj.query_len = ctx->preparsed_request_obj->uri_query.length;
+  }
+
+  /* Obtain handle to buffer containing the serialized payload */
+  const uint8_t *payload = NULL;
+  size_t payload_len = 0;
+#ifdef OC_BLOCK_WISE
+  if (*ctx->request_state) {
+    payload = (*ctx->request_state)->buffer;
+    payload_len = (*ctx->request_state)->payload_size;
   }
+#else  /* OC_BLOCK_WISE */
+  payload_len = coap_get_payload(ctx->request, &payload);
+#endif /* !OC_BLOCK_WISE */
+  request_obj._payload = payload;
+  request_obj._payload_len = payload_len;
+  request_obj.content_format = ctx->preparsed_request_obj->cf;
+  request_obj.accept = ctx->preparsed_request_obj->accept;
+#ifdef OC_HAS_FEATURE_ETAG
+  request_obj.etag = NULL;
+  request_obj.etag_len = coap_options_get_etag(ctx->request, &request_obj.etag);
+#endif /* OC_HAS_FEATURE_ETAG */
+
+  /* Attempt to locate the specific resource object that will handle the request
+   * using the request uri. */
+  oc_resource_t *cur_resource = ctx->preparsed_request_obj->cur_resource;
+  if (cur_resource != NULL) {
+    request_obj.resource = cur_resource;
+  }
+
+  bitmask_code_t bitmask_code = BITMASK_CODE_OK;
+  oc_interface_mask_t iface_mask = get_iface_mask(ctx->preparsed_request_obj);
+  if (!cur_resource) {
+    bitmask_code |= BITMASK_CODE_NOT_FOUND;
+  }
+
+#ifdef OC_HAS_FEATURE_ETAG
+  bitmask_code |= ri_invoke_coap_set_etag(
+    &(ri_invoke_coap_set_etag_in_t){
+      .request = ctx->request,
+      .preparsed_request_obj = ctx->preparsed_request_obj,
+      .bitmask_code = bitmask_code,
+      .endpoint = endpoint,
+      .method = method,
+      .iface_mask = iface_mask,
+    },
+    &request_obj);
+#endif /* OC_HAS_FEATURE_ETAG */
+
+  if (!bitmask_code && cur_resource != NULL) {
+    ri_invoke_handler_out_t out = {
+      .bitmask_code = bitmask_code,
+      .response_buffer = &response_buffer,
+    };
+    ri_invoke_handler(ctx,
+                      &(ri_invoke_handler_in_t){
+                        .request_obj = &request_obj,
+                        .preparsed_request_obj = ctx->preparsed_request_obj,
+                        .endpoint = endpoint,
+                        .method = method,
+                        .iface_mask = iface_mask,
+                        .payload = payload,
+                        .payload_len = payload_len,
+                      },
+                      &out);
+    bitmask_code |= out.bitmask_code;
```

Comment:
since we're ORing we can just return the added flags or 0 from `ri_invoke_handler`.

I would slightly prefer it that way since using output variable implies that you can remove flags from `out->bitmask_code` and that change would propagate here, but that's not true.
I guess we could do
```C
bitmask_code = out.bitmask_code
```
instead of OR-ing, but return value seems simpler to me

Reply:
fixed